### ISSUE TITLE
Windows -> Raw exec = Broken with 1.8.2

### ISSUE THREAD

**Issue by guifran001** [hashicorp/nomad#23668]
### Nomad version
Nomad v1.8.2
BuildDate 2024-07-16T08:50:09Z
Revision 7f0822c1e4f25907d9f60e2d595411950dd1bd28

### Operating system and Environment details

Windows
driver: raw_exec

### Issue

Jobs are stopping unexpectedly in nomad, but the processes are still alive in Windows, so when nomad tries to restart them, it fails because the port of application is already bound or a lock file is already lock.

It might not be 100% reproductible as we have an automated test that deploys a nomad instance and starts some jobs that never failed in 1.7.3, but now is failing since our update to 1.8.2. The test has been successful once with 1.8.2.

The automated test deploy consul 1.17.2 as both server and client, then nomad 1.8.2 as both server and client, then starts a postgres job and a keycloak job.

From the logs, we can see that postgres and keycloak start, then postgres stops and keycloak stops with Exit Code 2. Then they are stuck in a restart loop as the orignial postgres and keycloak processes are still executing.

From the log, we can see the Exit Code 2 that is usually reserved to file not found error. Could it be related to #23595 ?

### Reproduction steps

80%. The test is using pulumi to run the jobs. Each job is run sequentially

#### Expected Result

postgres and keycloak run fine

#### Actual Result

postgres and keycloak are stuck in a restart loop while a detached process of postgres and one from keycloak started by nomad are still running while nomad try to start new ones.

### Job file (if appropriate)

Will be provided if needed.

### Nomad Client logs (if appropriate)
```
POSTGRES Jobs is scheduled: 
2024-07-19T17:47:48.093Z [INFO]  client.alloc_runner.task_runner: Task event: alloc_id=79de797f-3e61-54a6-ce48-2affc507d636 task=postgres type=Received msg="Task received by client" failed=false
2024-07-19T17:47:48.110Z [INFO]  client.alloc_runner.task_runner: Task event: alloc_id=79de797f-3e61-54a6-ce48-2affc507d636 task=postgres type="Task Setup" msg="Building Task Directory" failed=false
2024-07-19T17:47:48.353Z [WARN]  client.alloc_runner.task_runner.task_hook.api: error creating task api socket: alloc_id=79de797f-3e61-54a6-ce48-2affc507d636 task=postgres path=c:\data\nomad-1.8.2\alloc\79de797f-3e61-54a6-ce48-2affc507d636\postgres\secrets\api.sock error="listen unix c:\data\nomad-1.8.2\alloc\79de797f-3e61-54a6-ce48-2affc507d636\postgres\secrets\api.sock: bind: invalid argument"
POSTGRES task starts
2024-07-19T17:47:48.385Z [INFO]  client.driver_mgr.raw_exec: starting task: driver=raw_exec driver_cfg="{Command:C:/install/00Postgres-v16.1.13-update-manifest-versions.1/pgsql/bin/postgres.exe Args:[-p 5432] OverrideCgroupV2: OverrideCgroupV1:map[] OOMScoreAdj:0}"
2024-07-19T17:47:48.597Z [INFO]  client.alloc_runner.task_runner: Task event: alloc_id=79de797f-3e61-54a6-ce48-2affc507d636 task=postgres type=Started msg="Task started by client" failed=false

KEYCLOAK Jobs is scheduled
2024-07-19T17:49:04.930Z [INFO]  client.alloc_runner.task_runner: Task event: alloc_id=876ad764-d20b-600a-adcb-8f22186401f7 task=compile-keycloak type=Received msg="Task received by client" failed=false
2024-07-19T17:49:04.931Z [INFO]  client.alloc_runner.task_runner: Task event: alloc_id=876ad764-d20b-600a-adcb-8f22186401f7 task=keycloak type=Received msg="Task received by client" failed=false
2024-07-19T17:49:04.948Z [INFO]  client.alloc_runner.task_runner: Task event: alloc_id=876ad764-d20b-600a-adcb-8f22186401f7 task=compile-keycloak type="Task Setup" msg="Building Task Directory" failed=false
2024-07-19T17:49:05.158Z [WARN]  client.alloc_runner.task_runner.task_hook.api: error creating task api socket: alloc_id=876ad764-d20b-600a-adcb-8f22186401f7 task=compile-keycloak path=c:\data\nomad-1.8.2\alloc\876ad764-d20b-600a-adcb-8f22186401f7\compile-keycloak\secrets\api.sock error="listen unix c:\data\nomad-1.8.2\alloc\876ad764-d20b-600a-adcb-8f22186401f7\compile-keycloak\secrets\api.sock: bind: invalid argument"
KEYCLOAK-COMPILE task starts (lifecycle {hook = "prestart"sidecar = false}
2024-07-19T17:49:05.177Z [INFO]  client.driver_mgr.raw_exec: starting task: driver=raw_exec driver_cfg="{Command:C:/install/openjdk-v21.0.2/bin/java.exe Args:[-Dkc.config.build-and-exit=true -Dprogram.name=keycloak -Xms64m -Xmx512m -XX:MetaspaceSize=96M -XX:MaxMetaspaceSize=256m -Dfile.encoding=UTF-8 -Dsun.stdout.encoding=UTF-8 -Dsun.err.encoding=UTF-8 -Dstdout.encoding=UTF-8 -Dstderr.encoding=UTF-8 -XX:+ExitOnOutOfMemoryError -Djava.security.egd=file:/dev/urandom -XX:+UseParallelGC -XX:MinHeapFreeRatio=10 -XX:MaxHeapFreeRatio=20 -XX:GCTimeRatio=4 -XX:AdaptiveSizePolicyWeight=90 -XX:FlightRecorderOptions=stackdepth=512 --add-opens=java.base/java.util=ALL-UNNAMED --add-opens=java.base/java.util.concurrent=ALL-UNNAMED --add-opens=java.base/java.security=ALL-UNNAMED -Dkc.home.dir=C:/install/keycloak-v23.0.6 -Djboss.server.config.dir=c:\data\nomad-1.8.2\alloc\876ad764-d20b-600a-adcb-8f22186401f7\compile-keycloak\local/conf -Dkeycloak.theme.dir=C:/install/keycloak-v23.0.6/themes -Djava.util.logging.manager=org.jboss.logmanager.LogManager -Dquarkus-log-max-startup-records=10000 -Dquarkus.datasource.jdbc.acquisition-timeout=2000 -cp C:/install/keycloak-v23.0.6/lib/quarkus-run.jar;C:/install/keycloak-v23.0.6/lib/**/* io.quarkus.bootstrap.runner.QuarkusEntryPoint build --db=postgres] OverrideCgroupV2: OverrideCgroupV1:map[] OOMScoreAdj:0}"
2024-07-19T17:49:05.404Z [INFO]  client.alloc_runner.task_runner: Task event: alloc_id=876ad764-d20b-600a-adcb-8f22186401f7 task=compile-keycloak type=Started msg="Task started by client" failed=false
KEYCLOAK-COMPILE task ends
2024-07-19T17:49:23.430Z [INFO]  client.alloc_runner.task_runner: Task event: alloc_id=876ad764-d20b-600a-adcb-8f22186401f7 task=compile-keycloak type=Terminated msg="Exit Code: 0" failed=false
2024-07-19T17:49:23.438Z [ERROR] client.driver_mgr.raw_exec: destroying executor failed: driver=raw_exec error="rpc error: code = Unknown desc = executor failed to find process: OpenProcess: The parameter is incorrect."
2024-07-19T17:49:23.464Z [INFO]  client.driver_mgr.raw_exec.executor: plugin process exited: alloc_id=876ad764-d20b-600a-adcb-8f22186401f7 driver=raw_exec task_name=compile-keycloak plugin=C:\install\nomad-1.8.2\bin\nomad.exe id=4688
2024-07-19T17:49:23.465Z [INFO]  client.alloc_runner.task_runner: not restarting task: alloc_id=876ad764-d20b-600a-adcb-8f22186401f7 task=compile-keycloak reason="Restart unnecessary as task terminated successfully"
2024-07-19T17:49:23.479Z [INFO]  client.alloc_runner.task_runner: Task event: alloc_id=876ad764-d20b-600a-adcb-8f22186401f7 task=keycloak type="Task Setup" msg="Building Task Directory" failed=false
2024-07-19T17:49:23.626Z [WARN]  client.alloc_runner.task_runner.task_hook.api: error creating task api socket: alloc_id=876ad764-d20b-600a-adcb-8f22186401f7 task=keycloak path=c:\data\nomad-1.8.2\alloc\876ad764-d20b-600a-adcb-8f22186401f7\keycloak\secrets\api.sock error="listen unix c:\data\nomad-1.8.2\alloc\876ad764-d20b-600a-adcb-8f22186401f7\keycloak\secrets\api.sock: bind: invalid argument"
KEYCLOAK task starts
2024-07-19T17:49:23.642Z [INFO]  client.driver_mgr.raw_exec: starting task: driver=raw_exec driver_cfg="{Command:C:/install/openjdk-v21.0.2/bin/java.exe Args:[-Dkc.config.built=true -Dprogram.name=keycloak -Xms64m -Xmx512m -XX:MetaspaceSize=96M -XX:MaxMetaspaceSize=256m -Dfile.encoding=UTF-8 -Dsun.stdout.encoding=UTF-8 -Dsun.err.encoding=UTF-8 -Dstdout.encoding=UTF-8 -Dstderr.encoding=UTF-8 -XX:+ExitOnOutOfMemoryError -Djava.security.egd=file:/dev/urandom -XX:+UseParallelGC -XX:MinHeapFreeRatio=10 -XX:MaxHeapFreeRatio=20 -XX:GCTimeRatio=4 -XX:AdaptiveSizePolicyWeight=90 -XX:FlightRecorderOptions=stackdepth=512 --add-opens=java.base/java.util=ALL-UNNAMED --add-opens=java.base/java.util.concurrent=ALL-UNNAMED --add-opens=java.base/java.security=ALL-UNNAMED -Dkc.home.dir=C:/install/keycloak-v23.0.6 -Djboss.server.config.dir=c:\data\nomad-1.8.2\alloc\876ad764-d20b-600a-adcb-8f22186401f7\keycloak\local/conf -Dkeycloak.theme.dir=C:/install/keycloak-v23.0.6/themes -Djava.util.logging.manager=org.jboss.logmanager.LogManager -Dquarkus-log-max-startup-records=10000 -Dquarkus.datasource.jdbc.acquisition-timeout=2000 -cp C:/install/keycloak-v23.0.6/lib/quarkus-run.jar;C:/install/keycloak-v23.0.6/lib/**/* io.quarkus.bootstrap.runner.QuarkusEntryPoint start --optimized --db=postgres --db-url jdbc:postgresql://127.0.0.1:5432/keycloak?user=iac&password=xxx OverrideCgroupV2: OverrideCgroupV1:map[] OOMScoreAdj:0}"
2024-07-19T17:49:23.761Z [INFO]  client.alloc_runner.task_runner: Task event: alloc_id=876ad764-d20b-600a-adcb-8f22186401f7 task=keycloak type=Started msg="Task started by client" failed=false

!!! POSTGRES stops with Exit code 0 ??? 
2024-07-19T17:50:51.551Z [INFO]  client.alloc_runner.task_runner: Task event: alloc_id=79de797f-3e61-54a6-ce48-2affc507d636 task=postgres type=Terminated msg="Exit Code: 0, Exit Message: \"executor: error waiting on process: rpc error: code = Unavailable desc = error reading from server: read tcp 127.0.0.1:50071->127.0.0.1:14000: wsarecv: An existing connection was forcibly closed by the remote host.\"" failed=false
2024-07-19T17:50:51.569Z [ERROR] client.driver_mgr.raw_exec: destroying executor failed: driver=raw_exec error="rpc error: code = Unavailable desc = connection error: desc = \"transport: error while dialing: dial tcp 127.0.0.1:14000: connectex: No connection could be made because the target machine actively refused it.\""

!!! KEYCLOAK stops with Exit code 0 ???
2024-07-19T17:50:51.589Z [INFO]  client.alloc_runner.task_runner: Task event: alloc_id=876ad764-d20b-600a-adcb-8f22186401f7 task=keycloak type=Terminated msg="Exit Code: 0, Exit Message: \"executor: error waiting on process: rpc error: code = Unavailable desc = error reading from server: read tcp 127.0.0.1:50272->127.0.0.1:14001: wsarecv: An existing connection was forcibly closed by the remote host.\"" failed=false
2024-07-19T17:50:51.619Z [ERROR] client.driver_mgr.raw_exec: destroying executor failed: driver=raw_exec error="rpc error: code = Unavailable desc = connection error: desc = \"transport: error while dialing: dial tcp 127.0.0.1:14001: connectex: No connection could be made because the target machine actively refused it.\""

!!! Executor de POSTGRES fails with Exit code 2 ??? (does 2 mean "The system cannot find the file specified"?)
2024-07-19T17:50:51.714Z [ERROR] client.driver_mgr.raw_exec.executor: plugin process exited: alloc_id=79de797f-3e61-54a6-ce48-2affc507d636 driver=raw_exec task_name=postgres plugin=C:\install\nomad-1.8.2\bin\nomad.exe id=2184 error="exit status 2"
2024-07-19T17:50:51.714Z [INFO]  client.alloc_runner.task_runner: restarting task: alloc_id=79de797f-3e61-54a6-ce48-2affc507d636 task=postgres reason="Restart within policy" delay=18.647683748s
2024-07-19T17:50:51.715Z [INFO]  client.alloc_runner.task_runner: Task event: alloc_id=79de797f-3e61-54a6-ce48-2affc507d636 task=postgres type=Restarting msg="Task restarting in 18.647683748s" failed=false

Executor de KEYCLOAK fails with Exit code 2 ??? (does 2 mean "The system cannot find the file specified"?)
2024-07-19T17:50:51.783Z [ERROR] client.driver_mgr.raw_exec.executor: plugin process exited: alloc_id=876ad764-d20b-600a-adcb-8f22186401f7 driver=raw_exec task_name=keycloak plugin=C:\install\nomad-1.8.2\bin\nomad.exe id=1568 error="exit status 2"
2024-07-19T17:50:51.783Z [INFO]  client.alloc_runner.task_runner: restarting task: alloc_id=876ad764-d20b-600a-adcb-8f22186401f7 task=keycloak reason="Restart within policy" delay=34.485843068s
```

**Comment by tgross** [comment]
Hi @guifran001! I can't really figure out why your postgres process exited, but it looks like the shutdown of the executor is the issue here. When the `compile-keycloak` pre-start task finishes, we see the following logs:

> 2024-07-19T17:49:23.430Z [INFO]  client.alloc_runner.task_runner: Task event: alloc_id=876ad764-d20b-600a-adcb-8f22186401f7 task=compile-keycloak type=Terminated msg="Exit Code: 0" failed=false
> 2024-07-19T17:49:23.438Z [ERROR] client.driver_mgr.raw_exec: destroying executor failed: driver=raw_exec error="rpc error: code = Unknown desc = executor failed to find process: OpenProcess: The parameter is incorrect."
> 2024-07-19T17:49:23.464Z [INFO]  client.driver_mgr.raw_exec.executor: plugin process exited: alloc_id=876ad764-d20b-600a-adcb-8f22186401f7 driver=raw_exec task_name=compile-keycloak plugin=C:\install\nomad-1.8.2\bin\nomad.exe id=4688

This error is coming from [`executor.go#L612`](https://github.com/hashicorp/nomad/blob/v1.8.2/drivers/shared/executor/executor.go#L612). To hit this code at all, the executor still has the process PID it's kept around in memory, but when we call `os.FindProcess` that hits this code in the Go stdlib, which is just a Windows syscall:

```go
func findProcess(pid int) (p *Process, err error) {
	const da = syscall.STANDARD_RIGHTS_READ |
		syscall.PROCESS_QUERY_INFORMATION | syscall.SYNCHRONIZE
	h, e := syscall.OpenProcess(da, false, uint32(pid))
	if e != nil {
		return nil, NewSyscallError("OpenProcess", e)
	}
	return newProcess(pid, uintptr(h)), nil
}
```

Is there any possibility that your task daemonizes? That is, does it double fork and then closes the first forking process to divorce the child from the grandparent (the Nomad executor in this case)? If not, then I'm not sure what's up... you're running Nomad as a privileged user, right?

> Could it be related to #23595 ?

That's an error trying to access Linux cgroups, so highly unlikely to be related.

**Comment by guifran001** [comment]
@tgross : thank you for the reply.

Our test is not starting Nomad as a privileged user but only as the `Network Service` User. 

We haven't taken the time to properly investigate yet. But we'll give it a try with a privileged user.

From `OpenProcess` function (processthreadsapi.h) [doc](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-openprocess):

> If the specified process is the System Idle Process (0x00000000), the function fails and the last error code is ERROR_INVALID_PARAMETER. If the specified process is the System process or one of the Client Server Run-Time Subsystem (CSRSS) processes, this function fails and the last error code is ERROR_ACCESS_DENIED because their access restrictions prevent user-level code from opening them.

So it it more likely than the `pid` is 0. From the code, I cannot figure out how this would happen and what's different from version 1.7.x about that.

**Comment by tgross** [comment]
I've found that Windows returns `ERROR_INVALID_PARAMETER` at the drop of a hat, but that's a good catch. The added bit of context is the sometimes-unfortunately decision that Go made to have "zero values" for uninitialized fields, even if they're value-fields. And guess what the zero-value of an PID (integer) is? So it could be that we're somehow no longer capturing the PID from the child process and so when we try to shutdown the executor we're passing it that uninitialized field.

I don't have an answer yet, but that gives me a place to start looking at least.

**Comment by tgross** [comment]
Hi @PayStreamArmitsteadJ. What's you're seeing isn't identical, but it sure smells similar. I'm not sure where you're getting this idea from:
> The executor failed due to not being able to make a HTTP call to a local HTTP server.

The executor isn't going to be making any HTTP requests or outbound connections of any kind. The executor is a [go-plugin](https://github.com/hashicorp/go-plugin) server. So the Nomad client agent spawns the executor process and then connects to that process via either a unix domain socket or a local TCP connection on Windows. 

When the client restarts (ex. after an upgrade), it reads its on-disk state to find tasks that should still be running, and then re-connects to the TCP server it expects to find there. If it can't reconnect, you'll see an error like the one you saw here:

> 2024-08-05T15:15:24.674+0100 [ERROR] client.driver_mgr.raw_exec: failed to reattach to executor: driver=raw_exec error="error creating rpc client for executor plugin: Reattachment process not found" task_id=5391c4b9-73e7-6943-8459-620ce812c13f/task/cae7b09d

At that point, the client says "ok this task must've failed while I was offline".  That's just something Nomad has to expect could happen, so it restarts the task (or not) according to the policy in the [`restart`](https://developer.hashicorp.com/nomad/docs/job-specification/restart) block.

But if the executor task is missing, there's no way for the client to have any idea that the task process is running or not. We're supposed to always kill the child processes when the executor dies, for any reason.

And then that's compounded by what you're seeing here:

> 2024-08-05T18:24:05.593+0100 [ERROR] client.driver_mgr.raw_exec: destroying executor failed: driver=raw_exec error="rpc error: code = Unavailable desc = connection error: desc = \"transport: error while dialing: dial tcp 127.0.0.1:14005: connectex: No connection could be made because the target machine actively refused it.\""

In this case, it looks like the executor's TCP server is unavailable. I suspect in this case the executor process is gone already so we're just trying to send a message over a connection where nobody's listening anymore. Which is more evidence that the executor processes are getting killed unexpectedly.

---

A question for both @guifran001 and @PayStreamArmitsteadJ: when these task processes are left running, are their executor processes still running as well, or are the executor processes also gone?

**Comment by PayStreamArmitsteadJ** [comment]
@tgross 

I have only experienced this issue once and it was affecting our production cluster, so I had to remedy the situation right away. Therefore, I do not know if the executor processes were still running. It seems like this was a one-off issue caused by the upgrade.

**Comment by tgross** [comment]
> I have only experienced this issue once and it was affecting our production cluster, so I had to remedy the situation right away. Therefore, I do not know if the executor processes were still running. It seems like this was a one-off issue caused by the upgrade.

Ok, that's information in itself. Thanks.

**Comment by ORuessel** [comment]
Hello everybody, 
i have the same situation. We upgraded the nomad cluster in two steps from 1.5.2 to 1.7.3 to 1.8.2. In our situation we have "long live" running jobs over one day, 
At the end of the day, we shut down and suspend in AWS the client nodes. On the next morning sometimes we get the issue, that on one different node the process is still running and not fully shutdown by nomad.

```
Aug 21, '24 11:43:28 +0200	Killing	Sent interrupt. Waiting 5s before force killing
Aug 21, '24 11:43:21 +0200	Restarting	Task restarting in 16s
Aug 21, '24 11:43:21 +0200	Terminated	Exit Code: 0, Exit Message: "executor: error waiting on process: rpc error: code = Unavailable desc = error reading from server: read tcp 127.0.0.1:64431->127.0.0.1:14000: wsarecv: An existing connection was forcibly closed by the remote host."
Aug 21, '24 11:43:19 +0200	Started	Task started by client
Aug 21, '24 11:43:01 +0200	Restarting	Task restarting in 18s
Aug 21, '24 11:43:01 +0200	Terminated	Exit Code: 0, Exit Message: "executor: error waiting on process: rpc error: code = Unavailable desc = error reading from server: read tcp 127.0.0.1:64411->127.0.0.1:14000: wsarecv: An existing connection was forcibly closed by the remote host."
Aug 21, '24 11:43:00 +0200	Started	Task started by client
Aug 21, '24 11:42:44 +0200	Restarting	Task restarting in 15s
Aug 21, '24 11:42:44 +0200	Terminated	Exit Code: 0, Exit Message: "executor: error waiting on process: rpc error: code = Unavailable desc = error reading from server: read tcp 127.0.0.1:64400->127.0.0.1:14000: wsarecv: An existing connection was forcibly closed by the remote host."
Aug 21, '24 11:42:43 +0200	Started	Task started by client
```

There was a old instance already running. After stop the process it was not possible to start the process in a normal way. only restart the hole machine was a temporary solution. The logfiles of nomad are crashed for today at this point i cannot provide at the moment and will monitor this situation and will deliver logs afterwards.

**Comment by tk2232** [comment]
I've encountered the same issue on Windows 10 and 11. 
After downgrading from Nomad 1.8.3 to 1.7.7, everything worked as expected again.

**Comment by tgross** [comment]
I've updated this issue to make sure it gets some attention on our internal tracker. I also have some suspicion that this may be related to https://github.com/hashicorp/nomad/issues/23969

**Comment by CullenShane** [comment]
I was able to reproduce this issue on Nomad 1.8.1 and 1.8.4 on Windows 11 with a minimal batch job. 

```
job "issue-23668" {
  datacenters = ["sally"]
  type = "batch"
  
  task "task" {
    driver = "raw_exec"

    config {
      command = "PowerShell"
      args = [
        "-Command",
        "Start-Sleep -Seconds 5"
      ]
    }
  }
} 
```

Failure logs:
```
2024-10-04T17:02:35.616-0500 [INFO]  client.alloc_runner.task_runner: Task event: alloc_id=0cbf4574-ebb0-b8d0-d6c5-140291560959 task=popup type=Received msg="Task received by client" failed=false
2024-10-04T17:02:35.630-0500 [INFO]  client.alloc_runner.task_runner: Task event: alloc_id=0cbf4574-ebb0-b8d0-d6c5-140291560959 task=popup type="Task Setup" msg="Building Task Directory" failed=false
2024-10-04T17:02:35.727-0500 [WARN]  client.alloc_runner.task_runner: some environment variables not available for rendering: alloc_id=0cbf4574-ebb0-b8d0-d6c5-140291560959 task=popup keys=""
2024-10-04T17:02:35.727-0500 [INFO]  client.driver_mgr.raw_exec: starting task: driver=raw_exec driver_cfg="{Command:PowerShell Args:[-Command Start-Sleep -Seconds 5] OverrideCgroupV2: OverrideCgroupV1:map[] OOMScoreAdj:0}"
2024-10-04T17:02:35.805-0500 [INFO]  client.alloc_runner.task_runner: Task event: alloc_id=0cbf4574-ebb0-b8d0-d6c5-140291560959 task=popup type=Started msg="Task started by client" failed=false
2024-10-04T17:02:37.433-0500 [INFO]  client.alloc_runner.task_runner: Task event: alloc_id=0cbf4574-ebb0-b8d0-d6c5-140291560959 task=popup type=Terminated msg="Exit Code: 0, Exit Message: \"executor: error waiting on process: rpc error: code = Unavailable desc = error reading from server: read tcp 127.0.0.1:59504->127.0.0.1:14000: wsarecv: An existing connection was forcibly closed by the remote host.\"" failed=false
2024-10-04T17:02:37.437-0500 [ERROR] client.driver_mgr.raw_exec: destroying executor failed: driver=raw_exec error="rpc error: code = Unavailable desc = connection error: desc = \"transport: error while dialing: dial tcp 127.0.0.1:14000: connectex: No connection could be made because the target machine actively refused it.\""
2024-10-04T17:02:37.530-0500 [ERROR] client.driver_mgr.raw_exec.executor: plugin process exited: alloc_id=0cbf4574-ebb0-b8d0-d6c5-140291560959 driver=raw_exec task_name=popup plugin=C:\Nomad\nomad.exe id=13416 error="exit status 2"
2024-10-04T17:02:37.530-0500 [INFO]  client.alloc_runner.task_runner: restarting task: alloc_id=0cbf4574-ebb0-b8d0-d6c5-140291560959 task=popup reason="Restart within policy" delay=17.91890335s
2024-10-04T17:02:37.530-0500 [INFO]  client.alloc_runner.task_runner: Task event: alloc_id=0cbf4574-ebb0-b8d0-d6c5-140291560959 task=popup type=Restarting msg="Task restarting in 17.91890335s" failed=false

```

Once I downgraded to 1.7.7 the issue went away.

Success on 1.7.7: 
```
    2024-10-04T17:05:16.995-0500 [INFO]  client.alloc_runner.task_runner: Task event: alloc_id=5dd66d9b-0ad6-5581-356b-d28f0e56a653 task=popup type=Received msg="Task received by client" failed=false
    2024-10-04T17:05:17.007-0500 [INFO]  client.alloc_runner.task_runner: Task event: alloc_id=5dd66d9b-0ad6-5581-356b-d28f0e56a653 task=popup type="Task Setup" msg="Building Task Directory" failed=false
    2024-10-04T17:05:17.105-0500 [WARN]  client.alloc_runner.task_runner: some environment variables not available for rendering: alloc_id=5dd66d9b-0ad6-5581-356b-d28f0e56a653 task=popup keys=""
    2024-10-04T17:05:17.107-0500 [INFO]  client.driver_mgr.raw_exec: starting task: driver=raw_exec driver_cfg="{Command:PowerShell Args:[-Command Start-Sleep -Seconds 5]}"
    2024-10-04T17:05:17.185-0500 [INFO]  client.alloc_runner.task_runner: Task event: alloc_id=5dd66d9b-0ad6-5581-356b-d28f0e56a653 task=popup type=Started msg="Task started by client" failed=false
    2024-10-04T17:05:22.478-0500 [INFO]  client.alloc_runner.task_runner: Task event: alloc_id=5dd66d9b-0ad6-5581-356b-d28f0e56a653 task=popup type=Terminated msg="Exit Code: 0" failed=false
    2024-10-04T17:05:22.490-0500 [INFO]  client.driver_mgr.raw_exec.executor: plugin process exited: alloc_id=5dd66d9b-0ad6-5581-356b-d28f0e56a653 driver=raw_exec task_name=popup plugin=C:\Nomad\nomad.exe id=10828
    2024-10-04T17:05:22.490-0500 [INFO]  client.alloc_runner.task_runner: not restarting task: alloc_id=5dd66d9b-0ad6-5581-356b-d28f0e56a653 task=popup reason="Restart unnecessary as task terminated successfully"
    2024-10-04T17:05:22.506-0500 [INFO]  client.alloc_runner.task_runner.task_hook.logmon: plugin process exited: alloc_id=5dd66d9b-0ad6-5581-356b-d28f0e56a653 task=popup plugin=C:\Nomad\nomad.exe id=8764
    2024-10-04T17:05:22.506-0500 [INFO]  client.gc: marking allocation for GC: alloc_id=5dd66d9b-0ad6-5581-356b-d28f0e56a653

```

**Comment by tgross** [comment]
@CullenShane I was hoping your reproduction would be the clue I needed, but I think it's possibly related but not _quite_ the same. It looks like we're simply returning an error spuriously. I ran that job and used process explorer to look at the process tree. The task was PID6000, and the `executor.out` log for the batch process reads like this:

```json
{"@level":"debug","@message":"plugin address","@timestamp":"2024-10-14T15:59:21.207440Z","address":"127.0.0.1:14000","network":"tcp"}
{"@level":"debug","@message":"shutdown requested","@module":"executor","@timestamp":"2024-10-14T16:00:21.488858Z","grace_period_ms":0,"signal":""}
{"@level":"warn","@message":"failed to shutdown due to inability to find process","@module":"executor","@timestamp":"2024-10-14T16:00:21.488858Z","error":"executor failed to find process: OpenProcess: The parameter is incorrect.","pid":6000}
```

At the point we're calling `Shutdown`, process 6000 has already exited. But of course it has! In theory there's nothing to do so we should be able to return `nil` rather than an error, but I'm worried doing so may be leaking child processes.

In the reports above the executor fails (the parent of the task itself), and then the task isn't stopping as it shuld. For folks who have seen this, it would be helpful if you could get the `executor.out` logs from the failed allocations.

**Comment by tgross** [comment]
Ok, I tried some more to reproduce this and I want to make sure that I've documented here a specific bug where the proximate cause is the executor process fails so that folks can report whether this is what they're seeing.

I'm running Nomad as a Windows service and deploying a web application `echo.exe` as a `raw_exec` service job:

```hcl
job "example" {

  task "task" {
    driver = "raw_exec"

    config {
      command = "C:/Users/Administrator/echo/echo.exe"
    }
  }
}
```

Killing the `echo.exe` process is uninteresting and the task restarts as expected.


Instead, let's look at what happens if we kill the executor. Before doing so, using [Process Explorer](https://learn.microsoft.com/en-us/sysinternals/downloads/process-explorer) I can see my process tree looks like this:

```
(1060) C:/Users/Administrator/nomad.exe agent -dev -config=C:/Users/Administrator/config.hcl
├── (3836) C:/Users/Administrator/nomad.exe logmon
│   └── (3740) C:/Windows/system32/conhost.exe 0x4
└── (3424) C:/Users/Administrator/nomad.exe executor
    ├── (4420) C:/Windows/system32/conhost.exe 0x4
    └── (2240) C:/Users/Administrator/echo/echo.exe
```


Our client logs look like the following. Note this tells us where to find the `executor.out` log (which can also be read via `nomad alloc fs`):

> 2024-10-14T18:33:59.355Z [INFO]  client.driver_mgr.raw_exec: starting task: driver=raw_exec driver_cfg="{Command:C:/Users/Administrator/echo/echo.exe Args:[] OverrideCgroupV2: OverrideCgroupV1:map[] OOMScoreAdj:0}"
> 2024-10-14T18:33:59.355Z [DEBUG] client.driver_mgr.raw_exec.executor: starting plugin: alloc_id=ac14131d-352a-d5b5-9270-6ec3dc0267bd driver=raw_exec task_name=task path=C:\Users\Administrator\nomad.exe args=["C:\\Users\\Administrator\\nomad.exe", "executor", "{\"LogFile\":\"C:\\\\Windows\\\\SystemTemp\\\\NomadClient183761340\\\\ac14131d-352a-d5b5-9270-6ec3dc0267bd\\\\task\\\\executor.out\",\"LogLevel\":\"debug\",\"FSIsolation\":false,\"Compute\":{\"tc\":4600,\"nc\":2}}"]
> 2024-10-14T18:33:59.358Z [DEBUG] client.driver_mgr.raw_exec.executor: plugin started: alloc_id=ac14131d-352a-d5b5-9270-6ec3dc0267bd driver=raw_exec task_name=task path=C:\Users\Administrator\nomad.exe pid=3424
> 2024-10-14T18:33:59.358Z [DEBUG] client.driver_mgr.raw_exec.executor: waiting for RPC address: alloc_id=ac14131d-352a-d5b5-9270-6ec3dc0267bd driver=raw_exec task_name=task plugin=C:\Users\Administrator\nomad.exe
> 2024-10-14T18:33:59.455Z [DEBUG] client.driver_mgr.raw_exec.executor: using plugin: alloc_id=ac14131d-352a-d5b5-9270-6ec3dc0267bd driver=raw_exec task_name=task version=2
> 2024-10-14T18:33:59.479Z [INFO]  client.alloc_runner.task_runner: Task event: alloc_id=ac14131d-352a-d5b5-9270-6ec3dc0267bd task=task type=Started msg="Task started by client" failed=false


Now kill the `echo.exe` process via Process Explorer and we'll see the following in our client logs:


> 2024-10-14T18:41:59.936Z [INFO]  client.alloc_runner.task_runner: Task event: alloc_id=ac14131d-352a-d5b5-9270-6ec3dc0267bd task=task type=Terminated msg="Exit Code: 0, Exit Message: \"executor: error waiting on process: rpc error: code = Unavailable desc = error reading from server: read tcp 127.0.0.1:50900->127.0.0.1:14000: wsarecv: An existing connection was forcibly closed by the remote host.\"" failed=false
> 2024-10-14T18:41:59.936Z [DEBUG] client.driver_mgr.raw_exec.executor.stdio: received EOF, stopping recv loop: alloc_id=ac14131d-352a-d5b5-9270-6ec3dc0267bd driver=raw_exec task_name=task err="rpc error: code = Unavailable desc = error reading from server: read tcp 127.0.0.1:50900->127.0.0.1:14000: wsarecv: An existing connection was forcibly closed by the remote host."
> 2024-10-14T18:41:59.936Z [ERROR] client.driver_mgr.raw_exec: destroying executor failed: driver=raw_exec error="rpc error: code = Unavailable desc = connection error: desc = \"transport: error while dialing: dial tcp 127.0.0.1:14000: connectex: No connection could be made because the target machine actively refused it.\""
> 2024-10-14T18:41:59.945Z [ERROR] client.driver_mgr.raw_exec.executor: plugin process exited: alloc_id=ac14131d-352a-d5b5-9270-6ec3dc0267bd driver=raw_exec task_name=task plugin=C:\Users\Administrator\nomad.exe id=3424 error="exit status 1"
> 2024-10-14T18:41:59.945Z [DEBUG] client.driver_mgr.raw_exec.executor: plugin exited: alloc_id=ac14131d-352a-d5b5-9270-6ec3dc0267bd driver=raw_exec task_name=task
> 2024-10-14T18:41:59.945Z [INFO]  client.alloc_runner.task_runner: restarting task: alloc_id=ac14131d-352a-d5b5-9270-6ec3dc0267bd task=task reason="Restart within policy" delay=15.884342135s

At this point, we can see in Process Explorer that the `echo.exe` process has been leaked. Because it's bound to a socket, the restart will fail. If we wait 15s for the restart, we see the following in the client logs:

> 2024-10-14T18:42:15.833Z [DEBUG] client.driver_mgr.raw_exec.executor: starting plugin: alloc_id=ac14131d-352a-d5b5-9270-6ec3dc0267bd driver=raw_exec task_name=task path=C:\Users\Administrator\nomad.exe args=["C:\\Users\\Administrator\\nomad.exe", "executor", "{\"LogFile\":\"C:\\\\Windows\\\\SystemTemp\\\\NomadClient183761340\\\\ac14131d-352a-d5b5-9270-6ec3dc0267bd\\\\task\\\\executor.out\",\"LogLevel\":\"debug\",\"FSIsolation\":false,\"Compute\":{\"tc\":4600,\"nc\":2}}"]
> 2024-10-14T18:42:15.837Z [DEBUG] client.driver_mgr.raw_exec.executor: plugin started: alloc_id=ac14131d-352a-d5b5-9270-6ec3dc0267bd driver=raw_exec task_name=task path=C:\Users\Administrator\nomad.exe pid=3236
> 2024-10-14T18:42:15.837Z [DEBUG] client.driver_mgr.raw_exec.executor: waiting for RPC address: alloc_id=ac14131d-352a-d5b5-9270-6ec3dc0267bd driver=raw_exec task_name=task plugin=C:\Users\Administrator\nomad.exe
> 2024-10-14T18:42:15.934Z [DEBUG] client.driver_mgr.raw_exec.executor: using plugin: alloc_id=ac14131d-352a-d5b5-9270-6ec3dc0267bd driver=raw_exec task_name=task version=2
> 2024-10-14T18:42:15.957Z [INFO]  client.alloc_runner.task_runner: Task event: alloc_id=ac14131d-352a-d5b5-9270-6ec3dc0267bd task=task type=Started msg="Task started by client" failed=false
> 2024-10-14T18:42:16.012Z [INFO]  client.alloc_runner.task_runner: Task event: alloc_id=ac14131d-352a-d5b5-9270-6ec3dc0267bd task=task type=Terminated msg="Exit Code: 1" failed=false
> 2024-10-14T18:42:16.012Z [ERROR] client.driver_mgr.raw_exec: destroying executor failed: driver=raw_exec error="rpc error: code = Unknown desc = executor failed to find process: OpenProcess: The parameter is incorrect."


The `executor.out` logs look like the following:

```json
{"@level":"debug","@message":"plugin address","@timestamp":"2024-10-14T18:33:59.452893Z","address":"127.0.0.1:14000","network":"tcp"}
{"@level":"debug","@message":"plugin address","@timestamp":"2024-10-14T18:42:15.932447Z","address":"127.0.0.1:14000","network":"tcp"}
{"@level":"debug","@message":"shutdown requested","@module":"executor","@timestamp":"2024-10-14T18:42:16.012855Z","grace_period_ms":0,"signal":""}
{"@level":"warn","@message":"failed to shutdown due to inability to find process","@module":"executor","@timestamp":"2024-10-14T18:42:16.012940Z","error":"executor failed to find process: OpenProcess: The parameter is incorrect.","pid":5580}
```

Note that there are no logs in the `executor.out` for when we kill the executor, and we get the "OpenProcess" error only when the task is restarted and fails. All that follows from the Task Events we see as well. Here the first Terminated event is from the killed executor -- the Nomad client records that the go-plugin connection to the executor has been lost. And then the subsequent Terminated events is from the application failing (as expected, because we've leaked the process bound to the port).

```
Recent Events:
Time                  Type            Description
2024-10-14T18:42:34Z  Not Restarting  Exceeded allowed attempts 2 in interval 30m0s and mode is "fail"
2024-10-14T18:42:34Z  Terminated      Exit Code: 1
2024-10-14T18:42:34Z  Started         Task started by client
2024-10-14T18:42:16Z  Restarting      Task restarting in 18.621692608s
2024-10-14T18:42:16Z  Terminated      Exit Code: 1
2024-10-14T18:42:15Z  Started         Task started by client
2024-10-14T18:41:59Z  Restarting      Task restarting in 15.884342135s
2024-10-14T18:41:59Z  Terminated      Exit Code: 0, Exit Message: "executor: error waiting on process: rpc error: code = Unavailable desc = error reading from server: read tcp 127.0.0.1:50900->127.0.0.1:14000: wsarecv: An existing connection was forcibly closed by the remote host."
2024-10-14T18:33:59Z  Started         Task started by client
2024-10-14T18:33:59Z  Task Setup      Building Task Directory
```

So my hypothesis is:
* The executor is failing ("wsarecv: An existing connection was forcibly closed by the remote host").
* Then the workload process is leaked.
* This prevents the new workload process from succeeding.
* The error messages we see in the logs from that are themselves weird and misleading ("executor failed to find process: OpenProcess: The parameter is incorrect").

This makes for 3 bugs in a trenchcoat:
* The executor should not be failing. However, I did notice while testing with 1.8.x that the executor's resource usage was climbing. This is not the case with a build that includes #24182. So that might be fixed already but it's hard to tell why we'd be seeing this otherwise. If folks have `executor.out` logs that contain stack traces that'd be new information.
* The workload process should not be leaked.
* The error handling reporting here is wrong.

If folks have any information from their environments that contradicts that hypothesis, I'd appreciate sharing it because that would be helpful to know.

**Comment by tgross** [comment]
I've got a fix up here for the executor orphaning the child processes: https://github.com/hashicorp/nomad/pull/24214. I'll open a new issue for improving the UX of the error messages.

**Comment by tgross** [comment]
Opened #24220 and #24219 as follow-ups from this issue.
